units/generate_tasks.go

```diff
@@ -107,34 +107,53 @@ func (j *generateTasksJob) Run(ctx context.Context) {
 	t, err := task.FindOneId(j.TaskID)
 	if err != nil {
 		j.AddError(errors.Wrapf(err, "problem finding task %s", j.TaskID))
+		return
 	}
 	if t == nil {
 		j.AddError(errors.Errorf("task %s does not exist", j.TaskID))
+		return
 	}
 
 	err = j.generate(ctx, t)
 	if err != nil && !adb.ResultsNotFound(err) {
+		t, err := task.FindOneId(j.TaskID)
```

Comment:
How come you're duplicating the logic on line 107?

Reply:
`j.generate` takes time. It's possible for another job to have raced while `j.generate ran`. Actually, as I think about it though, this can still race. I've pushed the logic down into `task.MarkGeneratedTasks` deleted `SetGenerateTasksError`, and made the query atomic.